lang/printer/src/types.rs

```diff
@@ -27,14 +27,24 @@ pub type Alloc<'a> = pretty::Arena<'a, Anno>;
 pub type Builder<'a> = pretty::DocBuilder<'a, Alloc<'a>, Anno>;
 
 /// Operator precedences
-pub type Precedence = u32;
+///
+/// This corresponds to the precedence specified by the parser grammar.
+#[derive(PartialOrd, Ord, PartialEq, Eq, Copy, Clone)]
+pub enum Precedence {
+    Exp,
+    NonLet,
+    Ops,
+    App,
+    Holes,
+    Atom,
+}
 
 /// We implement the `Print` trait for all types that we want to prettyprint.
 /// It is sufficient to implement either the `print` or the `print_prec` function, depending
 /// on whether you need information about operator precedences or not.
 pub trait Print {
     fn print<'a>(&'a self, cfg: &PrintCfg, alloc: &'a Alloc<'a>) -> Builder<'a> {
-        Print::print_prec(self, cfg, alloc, 0)
+        Print::print_prec(self, cfg, alloc, Precedence::Exp)
```

Comment:
I think it is better to define a method on `Precedence`, something like `Precedence::highest()` and to annotate this function with a warning that it is only to be called on the outermost level, not for printing subexpressions.